File: ports/webassembly/jsfetch/jsfetch.c
// _jsfetch: a blocking HTTP primitive backed by the browser's fetch(). A
// browser can't open raw sockets, so the socket-based `requests` /
// `urllib.urequest` from micropython-lib don't work; this gives Python code a
// fetch-backed alternative to build on.
//
// request() kicks off the fetch then suspends the WASM stack (via
// emscripten_sleep) until it settles, giving Python an ordinary blocking call.
// It therefore requires a suspend-capable build (a JSPI or Asyncify variant);
// enable it by including jsfetch/jsfetch.mk, which sets MICROPY_PY_JSFETCH.

#include "py/runtime.h"
#include "py/objstr.h"

#if MICROPY_PY_JSFETCH

#include <emscripten.h>
#include <stdlib.h>
#include <string.h>

// Implemented in jsfetch/jsfetch.js.
extern void mp_js_fetch_start(const char *method, const char *url, const char *headers, const char *body, int body_len);
extern int mp_js_fetch_poll(void);     // 0 = pending, 1 = done, 2 = error
extern int mp_js_fetch_status(void);
extern int mp_js_fetch_body_len(void);
extern void mp_js_fetch_body_copy(void *dest);
extern char *mp_js_fetch_str(int which);  // 0 reason, 1 headers, 2 error (malloc'd)

// Streaming variant (simulator/jsfetch.js).
extern void mp_js_sfetch_start(const char *method, const char *url, const char *headers, const char *body, int body_len);
extern int mp_js_sfetch_phase(void);   // 0 = pending, 1 = response received, 3 = error
extern int mp_js_sfetch_status(void);
extern char *mp_js_sfetch_str(int which);  // 0 reason, 1 headers, 2 error (malloc'd)
extern int mp_js_sfetch_done(void);
extern int mp_js_sfetch_read(void *dest, int maxlen);

// request(method, url, headers="", body=None) -> (status, reason, headers, body)
//   headers: "Key: Value" lines separated by "\n"
//   body:    a bytes-like object, or None
//   returns: status (int), reason (str), response headers (str), body (bytes)
static mp_obj_t jsfetch_request(size_t n_args, const mp_obj_t *args) {
    const char *method = mp_obj_str_get_str(args[0]);
    const char *url = mp_obj_str_get_str(args[1]);
    const char *headers = (n_args > 2 && args[2] != mp_const_none) ? mp_obj_str_get_str(args[2]) : "";

    const char *body = "";
    int body_len = 0;
    if (n_args > 3 && args[3] != mp_const_none) {
        mp_buffer_info_t bi;
        mp_get_buffer_raise(args[3], &bi, MP_BUFFER_READ);
        body = (const char *)bi.buf;
        body_len = (int)bi.len;
    }

    mp_js_fetch_start(method, url, headers, body, body_len);
    while (mp_js_fetch_poll() == 0) {
        emscripten_sleep(5);  // JSPI-suspend; the fetch resolves on the event loop
    }

    if (mp_js_fetch_poll() == 2) {
        char *err = mp_js_fetch_str(2);
        mp_obj_t exc = mp_obj_new_exception_msg_varg(&mp_type_OSError, MP_ERROR_TEXT("fetch failed: %s"), err);
        free(err);
        nlr_raise(exc);
    }

    int status = mp_js_fetch_status();
    char *reason = mp_js_fetch_str(0);
    char *hdrs = mp_js_fetch_str(1);

    int blen = mp_js_fetch_body_len();
    vstr_t vstr;
    vstr_init_len(&vstr, blen);
    if (blen > 0) {
        mp_js_fetch_body_copy((void *)vstr.buf);
    }

    mp_obj_t tuple[4];
    tuple[0] = mp_obj_new_int(status);
    tuple[1] = mp_obj_new_str(reason, strlen(reason));
    tuple[2] = mp_obj_new_str(hdrs, strlen(hdrs));
    tuple[3] = mp_obj_new_bytes_from_vstr(&vstr);
    free(reason);
    free(hdrs);
    return mp_obj_new_tuple(4, tuple);
}
static MP_DEFINE_CONST_FUN_OBJ_VAR_BETWEEN(jsfetch_request_obj, 2, 4, jsfetch_request);

// ---- streaming API, for a cooperative AsyncFetch-style client --------------

// stream_start(method, url, headers="", body=None): kick off a streaming fetch.
static mp_obj_t jsfetch_stream_start(size_t n_args, const mp_obj_t *args) {
    const char *method = mp_obj_str_get_str(args[0]);
    const char *url = mp_obj_str_get_str(args[1]);
    const char *headers = (n_args > 2 && args[2] != mp_const_none) ? mp_obj_str_get_str(args[2]) : "";

    const char *body = "";
    int body_len = 0;
    if (n_args > 3 && args[3] != mp_const_none) {
        mp_buffer_info_t bi;
        mp_get_buffer_raise(args[3], &bi, MP_BUFFER_READ);
        body = (const char *)bi.buf;
        body_len = (int)bi.len;
    }
    mp_js_sfetch_start(method, url, headers, body, body_len);
    return mp_const_none;
}
static MP_DEFINE_CONST_FUN_OBJ_VAR_BETWEEN(jsfetch_stream_start_obj, 2, 4, jsfetch_stream_start);

// stream_phase() -> 0 pending, 1 response received (streaming), 3 error.
static mp_obj_t jsfetch_stream_phase(void) {
    return MP_OBJ_NEW_SMALL_INT(mp_js_sfetch_phase());
}
static MP_DEFINE_CONST_FUN_OBJ_0(jsfetch_stream_phase_obj, jsfetch_stream_phase);

// stream_response() -> (status:int, reason:str, headers:str)
static mp_obj_t jsfetch_stream_response(void) {
    char *reason = mp_js_sfetch_str(0);
    char *hdrs = mp_js_sfetch_str(1);
    mp_obj_t tuple[3];
    tuple[0] = MP_OBJ_NEW_SMALL_INT(mp_js_sfetch_status());
    tuple[1] = mp_obj_new_str(reason, strlen(reason));
    tuple[2] = mp_obj_new_str(hdrs, strlen(hdrs));
    free(reason);
    free(hdrs);
    return mp_obj_new_tuple(3, tuple);
}
static MP_DEFINE_CONST_FUN_OBJ_0(jsfetch_stream_response_obj, jsfetch_stream_response);

// stream_error() -> str
static mp_obj_t jsfetch_stream_error(void) {
    char *err = mp_js_sfetch_str(2);
    mp_obj_t s = mp_obj_new_str(err, strlen(err));
    free(err);
    return s;
}
static MP_DEFINE_CONST_FUN_OBJ_0(jsfetch_stream_error_obj, jsfetch_stream_error);

// stream_done() -> bool (body fully read and all buffered bytes consumed)
static mp_obj_t jsfetch_stream_done(void) {
    return mp_obj_new_bool(mp_js_sfetch_done());
}
static MP_DEFINE_CONST_FUN_OBJ_0(jsfetch_stream_done_obj, jsfetch_stream_done);

// stream_readinto(buf) -> int bytes copied (0 if none buffered yet)
static mp_obj_t jsfetch_stream_readinto(mp_obj_t buf_in) {
    mp_buffer_info_t bi;
    mp_get_buffer_raise(buf_in, &bi, MP_BUFFER_WRITE);
    return MP_OBJ_NEW_SMALL_INT(mp_js_sfetch_read(bi.buf, (int)bi.len));
}
static MP_DEFINE_CONST_FUN_OBJ_1(jsfetch_stream_readinto_obj, jsfetch_stream_readinto);

static const mp_rom_map_elem_t jsfetch_globals_table[] = {
    { MP_ROM_QSTR(MP_QSTR___name__), MP_ROM_QSTR(MP_QSTR__jsfetch) },
    { MP_ROM_QSTR(MP_QSTR_request), MP_ROM_PTR(&jsfetch_request_obj) },
    { MP_ROM_QSTR(MP_QSTR_stream_start), MP_ROM_PTR(&jsfetch_stream_start_obj) },
    { MP_ROM_QSTR(MP_QSTR_stream_phase), MP_ROM_PTR(&jsfetch_stream_phase_obj) },
    { MP_ROM_QSTR(MP_QSTR_stream_response), MP_ROM_PTR(&jsfetch_stream_response_obj) },
    { MP_ROM_QSTR(MP_QSTR_stream_error), MP_ROM_PTR(&jsfetch_stream_error_obj) },
    { MP_ROM_QSTR(MP_QSTR_stream_done), MP_ROM_PTR(&jsfetch_stream_done_obj) },
    { MP_ROM_QSTR(MP_QSTR_stream_readinto), MP_ROM_PTR(&jsfetch_stream_readinto_obj) },
};
static MP_DEFINE_CONST_DICT(jsfetch_globals, jsfetch_globals_table);

const mp_obj_module_t mp_module__jsfetch = {
    .base = { &mp_type_module },
    .globals = (mp_obj_dict_t *)&jsfetch_globals,
};

MP_REGISTER_MODULE(MP_QSTR__jsfetch, mp_module__jsfetch);

#endif // MICROPY_PY_JSFETCH

File: ports/webassembly/jsfetch/jsfetch.js
// JavaScript half of the fetch-backed HTTP primitive (see jsfetch/jsfetch.c).
// Lives in a --js-library so it can use full JS (fetch, promises, object
// literals) without the EM_ASM comma/macro pitfalls.
//
// A single in-flight request is tracked in __mpFetch. The C side kicks it off
// with mp_js_fetch_start(), then polls mp_js_fetch_poll() while suspending the
// WASM stack until the promise settles, giving Python a blocking call.

mergeInto(LibraryManager.library, {
    mp_js_fetch_start__postset: "var __mpFetch = null;",

    // method/url/headers are NUL-terminated UTF-8; headers are "K: V" lines
    // separated by "\n". body is a pointer/length (length 0 means no body).
    mp_js_fetch_start: function (methodPtr, urlPtr, headersPtr, bodyPtr, bodyLen) {
        var method = UTF8ToString(methodPtr);
        var url = UTF8ToString(urlPtr);
        var headerStr = UTF8ToString(headersPtr);

        var state = { done: 0, status: 0, reason: "", headers: "", body: new Uint8Array(0), error: "" };
        __mpFetch = state;

        var opts = { method: method };
        if (headerStr) {
            var h = {};
            headerStr.split("\n").forEach(function (line) {
                var i = line.indexOf(":");
                if (i > 0) { h[line.slice(0, i).trim()] = line.slice(i + 1).trim(); }
            });
            opts.headers = h;
        }
        if (bodyLen > 0) {
            // Copy out of the WASM heap; the C buffer may be gone by the time
            // fetch reads it.
            opts.body = HEAPU8.slice(bodyPtr, bodyPtr + bodyLen);
        }

        fetch(url, opts).then(function (r) {
            state.status = r.status;
            state.reason = r.statusText || "";
            var hs = [];
            r.headers.forEach(function (v, k) { hs.push(k + ": " + v); });
            state.headers = hs.join("\r\n");
            return r.arrayBuffer();
        }).then(function (buf) {
            state.body = new Uint8Array(buf);
            state.done = 1;
        }).catch(function (e) {
            state.error = (e && e.message) ? e.message : ("" + e);
            state.done = 2;
        });
    },

    // 0 = pending, 1 = done, 2 = error.
    mp_js_fetch_poll: function () { return __mpFetch ? __mpFetch.done : 2; },

    mp_js_fetch_status: function () { return __mpFetch ? __mpFetch.status : 0; },

    mp_js_fetch_body_len: function () { return __mpFetch ? __mpFetch.body.length : 0; },

    mp_js_fetch_body_copy: function (dest) { if (__mpFetch) { HEAPU8.set(__mpFetch.body, dest); } },

    // which: 0 = reason, 1 = headers, 2 = error. Returns a malloc'd UTF-8 string
    // the caller must free().
    mp_js_fetch_str: function (which) {
        var s = "";
        if (__mpFetch) { s = which === 0 ? __mpFetch.reason : which === 1 ? __mpFetch.headers : __mpFetch.error; }
        return stringToNewUTF8(s || "");
    },

    // ---- streaming variant, for modules/fetch.py (AsyncFetch) ----------------
    // The body is read incrementally via a ReadableStream reader and queued so
    // the Python side can pull chunks across successive update() calls.
    mp_js_sfetch_start__postset: "var __mpSFetch = null;",

    mp_js_sfetch_start: function (methodPtr, urlPtr, headersPtr, bodyPtr, bodyLen) {
        var method = UTF8ToString(methodPtr);
        var url = UTF8ToString(urlPtr);
        var headerStr = UTF8ToString(headersPtr);

        // phase: 0 pending, 1 response received (streaming), 3 error.
        var st = { phase: 0, status: 0, reason: "", headers: "", chunks: [], avail: 0, readDone: false, error: "" };
        __mpSFetch = st;

        var opts = { method: method };
        if (headerStr) {
            var h = {};
            headerStr.split("\n").forEach(function (line) {
                var i = line.indexOf(":");
                if (i > 0) { h[line.slice(0, i).trim()] = line.slice(i + 1).trim(); }
            });
            opts.headers = h;
        }
        if (bodyLen > 0) { opts.body = HEAPU8.slice(bodyPtr, bodyPtr + bodyLen); }

        fetch(url, opts).then(function (r) {
            st.status = r.status;
            st.reason = r.statusText || "";
            var hs = [];
            r.headers.forEach(function (v, k) { hs.push(k + ": " + v); });
            st.headers = hs.join("\r\n");
            st.phase = 1;
            if (!r.body) { st.readDone = true; return; }
            var reader = r.body.getReader();
            var pump = function () {
                reader.read().then(function (res) {
                    if (res.done) { st.readDone = true; return; }
                    st.chunks.push(res.value);
                    st.avail += res.value.length;
                    pump();
                }).catch(function (e) {
                    st.error = (e && e.message) ? e.message : ("" + e);
                    st.readDone = true;
                });
            };
            pump();
        }).catch(function (e) {
            st.error = (e && e.message) ? e.message : ("" + e);
            st.phase = 3;
        });
    },

    mp_js_sfetch_phase: function () { return __mpSFetch ? __mpSFetch.phase : 3; },
    mp_js_sfetch_status: function () { return __mpSFetch ? __mpSFetch.status : 0; },

    // 0 = reason, 1 = headers, 2 = error. Returns a malloc'd UTF-8 string.
    mp_js_sfetch_str: function (which) {
        var s = "";
        if (__mpSFetch) { s = which === 0 ? __mpSFetch.reason : which === 1 ? __mpSFetch.headers : __mpSFetch.error; }
        return stringToNewUTF8(s || "");
    },

    // Done = body fully read AND everything buffered has been consumed.
    mp_js_sfetch_done: function () {
        if (!__mpSFetch) { return 1; }
        return (__mpSFetch.readDone && __mpSFetch.avail === 0) ? 1 : 0;
    },

    // Copy up to maxlen buffered bytes into dest; returns the count (0 if none
    // are available yet).
    mp_js_sfetch_read: function (dest, maxlen) {
        var st = __mpSFetch;
        if (!st || st.avail === 0) { return 0; }
        var written = 0;
        while (written < maxlen && st.chunks.length > 0) {
            var head = st.chunks[0];
            var take = Math.min(head.length, maxlen - written);
            HEAPU8.set(head.subarray(0, take), dest + written);
            written += take;
            if (take === head.length) { st.chunks.shift(); }
            else { st.chunks[0] = head.subarray(take); }
        }
        st.avail -= written;
        return written;
    },
});
